hazelcast/src/main/java/com/hazelcast/map/impl/MapEntries.java

```diff
@@ -85,4 +77,28 @@ public int getFactoryId() {
     public int getId() {
         return MapDataSerializerHook.ENTRY_SET;
     }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        int size = entries.size();
+        out.writeInt(size);
+
+        if (size > 0) {
+            for (Map.Entry<Data, Data> o : entries) {
+                out.writeData(o.getKey());
+                out.writeData(o.getValue());
+            }
+        }
+    }
+
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        int size = in.readInt();
+        for (int i = 0; i < size; i++) {
```

Comment:
it is a pain. I'd do something like this:
```java
private static final int ARRAY_RECREATION_THRESHOLD = 200;
[...]
public void readData() {
  [...]
  if (size > ARRAY_RECREATION_THRESHOLD) {
    assert entries.isEmpty()
    entries = new ArrayList<>(size);
  }
  [...]
}
```

It's ugly, the threshold might seems artificial, but I *believe* it will work better. The ugly code is isolated. Imagine you have extremely large result site. with 10,000's of entries.